src/common/base/src/runtime/allocator.rs

```diff
@@ -0,0 +1,411 @@
+        });
+
+        let (new_layout, offset) = Layout::new::<AllocHeader>().extend(layout).unwrap();
+
+        let res = self.inner.alloc(new_layout);
+        *res.cast::<AllocHeader>() = header;
+
+        res.add(offset)
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        let (new_layout, offset) = Layout::new::<AllocHeader>().extend(layout).unwrap();
+
+        let ptr = ptr.sub(offset);
+
+        let ah = &mut (*(ptr.cast::<AllocHeader>()));
+        debug_assert!(ah.is_allocated());
+        ah.mark_as_freed();
+        let header_tid = ah.tid;
+
+        MEM_SIZE[header_tid % COUNTERS_SIZE].fetch_sub(layout.size(), Ordering::Relaxed);
+        MEM_CNT[header_tid % COUNTERS_SIZE].fetch_sub(1, Ordering::Relaxed);
+
+        self.inner.dealloc(ptr, new_layout);
+    }
+}
+
+impl<A: GlobalAlloc> ProxyAllocator<A> {
+    unsafe fn print_stack_trace_on_memory_spike(layout: Layout, tid: usize, memory_usage: usize) {
+        MEMORY_USAGE_LAST_REPORT.with(|memory_usage_last_report| {
+            if memory_usage
+                > REPORT_USAGE_INTERVAL
+                    .load(Ordering::Relaxed)
+                    .saturating_add(memory_usage_last_report.get())
+            {
+                memory_usage_last_report.set(memory_usage);
+                tracing::warn!(
+                    tid,
+                    memory_usage_mb = memory_usage / MEBIBYTE,
+                    added_mb = layout.size() / MEBIBYTE,
+                    bt = ?Backtrace::new(),
+                    "reached new record of memory usage",
+                );
+            }
+        });
+    }
+}
+
+impl<A: GlobalAlloc> ProxyAllocator<A> {
+    #[inline]
+    unsafe fn compute_stack_trace(
+        layout: Layout,
+        stack: &mut [*mut c_void; STACK_SIZE],
+        verbose: bool,
+    ) {
+        if Self::should_compute_trace(layout) {
+            const MISSING_TRACE: *mut c_void = 2 as *mut c_void;
+            stack[0] = MISSING_TRACE;
+            backtrace::trace(|frame| {
+                let addr = frame.ip();
+                stack[0] = addr.cast::<c_void>();
+                if addr >= SKIP_ADDR_ABOVE.cast::<c_void>() {
+                    true
+                } else {
+                    let hash = (murmur64(addr as u64) % (8 * CACHE_SIZE as u64)) as usize;
+                    let i = hash / 8;
+                    let cur_bit = 1 << (hash % 8);
+                    if SKIP_CACHE[i] & cur_bit != 0 {
+                        true
+                    } else if CHECKED_CACHE[i] & cur_bit != 0 {
+                        false
+                    } else if skip_ptr(addr) {
+                        SKIP_CACHE[i] |= cur_bit;
+                        true
+                    } else {
+                        CHECKED_CACHE[i] |= cur_bit;
+                        false
+                    }
+                }
+            });
+            if verbose {
+                tracing::info!(?stack, "STARTED_TRACE");
```

Comment:
This call will call alloc again, we should avoid call tracing in allocator.

Reply:
PoC phase, will remove them all when it's ready.

Reply:
> PoC phase, will remove them all when it's ready.

I'm guessing we will get `SIGABRT` here.